[Docs] macOS Dev environment installation

farhangnaderi opened the pull request:
### Solved Problem

Adding Docs for new gazebo setup and Dev env install on macOS
### Things we might need to figure out:

1. If Installation guide Video needs to stay still. 
2. Relevant issue #25205 

Fixes #25071

farhangnaderi committed:
Initial changes

farhangnaderi committed:
index fix

farhangnaderi committed:
gz index fix2

farhangnaderi committed:
gz index fix 3

mrpollo commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -72,9 +55,67 @@ To setup the environment to be able to build for Pixhawk/NuttX hardware (and ins
+
+   ```sh
+   brew install gz-harmonic
+   brew install qt@5 gstreamer gst-plugins-base gst-plugins-good gst-plugins-bad gst-plugins-ugly gst-libav
````
does the gz-harmonic formula not install its dependencies?

hamishwillee replied:
Why are we doing this if `./PX4-Autopilot/Tools/setup/macos.sh --sim-tools` sets everything up?

mrpollo commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -72,9 +55,67 @@ To setup the environment to be able to build for Pixhawk/NuttX hardware (and ins
+   export DYLD_LIBRARY_PATH="/opt/homebrew/lib:$DYLD_LIBRARY_PATH"
+   ```
+
+1. Link Qt5 and reload your shell configuration:
````
I recall that we had this discussion before. Although this can address a few initial issues, it's essentially a "magic" solution that isn't fully explained, and we shouldn't recommend it without also explaining its necessity. Let's find a way to be more explicit about it.

mrpollo commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -72,9 +55,67 @@ To setup the environment to be able to build for Pixhawk/NuttX hardware (and ins
+   brew install qt@5 gstreamer gst-plugins-base gst-plugins-good gst-plugins-bad gst-plugins-ugly gst-libav
+   ```
+
+1. Configure Qt5 and library paths by adding the following to your `~/.zshrc` file:
````
We need to reword this. You're not configuring Qt5; we're ensuring it's accessible to the default shell, which can be anything. The majority of users will be using `~/.bashrc`; let's default to this. Any Zsh users will know what to do from here.

The homebrew Qt5 formula gives you a very verbose explanation of what needs to happen after installation. We need to make it clear this isn't specific to PX4.

mrpollo commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -72,9 +55,67 @@ To setup the environment to be able to build for Pixhawk/NuttX hardware (and ins
 
-To setup the environment for [Gazebo Classic](../sim_gazebo_classic/index.md) simulation:
+   ```sh
+   ls -la /opt/homebrew/lib/cmake/Qt5Core/
````
Why is this step necessary? You already installed Qt5 twice above; this seems redundant. What's the issue we are trying to solve?

mrpollo reviewed (changes requested):
almost there thanks for putting the work @farhangnaderi

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -21,17 +21,6 @@ The "base" macOS setup installs the tools needed for building firmware, and incl
 
 ### Environment Setup
 
-:::details Apple Silicon Macbook users!
```
Does this means that the standard setup works for all types of silicon?

mrpollo replied:
yes it works for all!

hamishwillee commented:
@farhangnaderi Got an ETA on retesting using the script?

farhangnaderi commented:
I will bring this PR to life in a day or two.

farhangnaderi commented:
> @farhangnaderi I made changes to the [macOS installer](https://github.com/PX4/PX4-Autopilot/pull/25361) script and the [homebrew formula](https://github.com/PX4/homebrew-px4/pull/93) to address inconsistencies and fix CI for everyone. Please help document this properly. Note the main changes are as follows:
> 
> * The arm cross compiler used in the macOS environment was bumped from v9 to v13, which is now in line with what we are recommending for Ubuntu 24.04 (latest)
>   
>   * Installing the binaries is easy; the problem comes from Homebrew's reluctance to link the binaries. We have to add a manual step to force them. This is explained when you install the `px4-dev` formula via Homebrew with a '**caveat**' warning (see below)
> 
> ```
> ==> Caveats
> The PX4 development toolchain has been installed, including the ARM cross-compiler
> from the osx-cross/arm tap (arm-gcc-bin@13).
> 
> Homebrew does not link versioned formulae by default, so you must link the compiler manually.
> 
>     brew link --overwrite --force arm-gcc-bin@13
> 
> If you have other versions of arm-none-eabi-gcc installed, this may override them.
> You can unlink it manually with:
> 
>     brew unlink arm-gcc-bin@13
> ```
> 
>  * We need to replace all of the simulation steps to use the `px4-sim` formula, which was carefully crafted to install all dependencies. If you test this and it doesn't work, then we have to fix the issue instead of doing workarounds.
> * The docs should be as simple as `brew install px4-dev px4-sim` except the link step above; everything else should be removed unless it's been proven to be needed. Assume it's old info that is no longer valid.

I need to try this on a clean env and update the docs. 
thanks

farhangnaderi committed:
updates

farhangnaderi commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -72,9 +55,67 @@ To setup the environment to be able to build for Pixhawk/NuttX hardware (and ins
+   sh macos.sh
+   ```
+
+## Unsupported Simulation Options
+
+### Gazebo Classic
````
same question here.

hamishwillee committed:
Run prettier

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -1,40 +1,27 @@
+1. Install Homebrew by following these [installation instructions](https://brew.sh).
 
-1. Enable more open files by appending the following line to the `~/.zshenv` file (creating it if necessary):
+2. Enable more open files by appending the following line to the `~/.zshenv` file (creating it if necessary):
```
I don't use .zshenv. Is this the default thing that everyone uses for all shells on macos?

hamishwillee reviewed (commented):
@mrpollo I can't test this, but it looks sane/good. Certainly much better than before if it works because the instructions are simple. 

1. Should we strip out the gazebo classic section.
2. @farhangnaderi Do other macOS sections in docs work too - e.g. https://docs.px4.io/main/en/sim_jmavsim/#macos

farhangnaderi committed:
zsh env
removed gz classic

farhangnaderi committed:
Corrections 2

hamishwillee reviewed (dismissed):
THanks @farhangnaderi - I like this. I'm happy for it to merge and any jmavsim testing to follow - after all, that is "community supported". I don't want to forget about it though!

@mrpollo Approving but not merging, because I can't test this.

mrpollo commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -1,112 +1,83 @@
 
    ```sh
-   # Point pip3 to MacOS system python 3 pip
+   # Point pip3 to macOS system python 3 pip
````
Since we are rewriting most of the docs already, why not update this section too:

I think this part is misguided, although it might work for some users, it depends on the base os level Python3 existing and the user not having installed a later version of Python3 via package manager or otherwise.

I suggest we say "Ensure Python 3 is used by default, since some scripts found in PX4 require python3 and pip3 executables in your path.

Then afterwards we can have a TIP that gives users a recommended way to install pyhon 3, which should say something like this

```
::: tip
If you need help installing an updated version of Python 3, we recommend the [pyenv project](https://github.com/pyenv/pyenv?tab=readme-ov-file), which gives you utmost flexibility by allowing you to set a global and local Python version at the per-directory level.
```

mengchaoheng replied:
Before `brew install px4-dev px4-sim`:

```bash
mch@mchdexuniji ~ % which python3
/usr/bin/python3
mch@mchdexuniji ~ % which pip
pip not found
mch@mchdexuniji ~ % which pip3
pip3: aliased to /usr/bin/pip3

```
After running `brew install px4-dev px4-sim`:
```bash
mch@mchdexuniji ~ % which python3
/opt/homebrew/bin/python3
mch@mchdexuniji ~ % which pip3   
pip3: aliased to /usr/bin/pip3
mch@mchdexuniji ~ % brew list
==> Formulae
ant			gettext			lzo
arm-gcc-bin@13		giflib			m4
asio@1.10.8		glib			mpdecimal
astyle			graphite2		ncurses
autoconf		harfbuzz		ninja
automake		hiredis			openjdk
bash-completion		icu4c@77		openssl@3
blake3			jpeg-turbo		pcre2
boost@1.85		kconfig-frontends	pixman
ca-certificates		libpng			pkgconf
cairo			libtiff			px4-dev
ccache			libtommath		python-tk@3.13
cmake			libtool			python@3.13
expat			libunistring		readline
fastcdr			libx11			sqlite
fastdds			libxau			tcl-tk
flock			libxcb			tinyxml2
fmt			libxdmcp		xorgproto
fontconfig		libxext			xxhash
foonathan-memory	libxrender		xz
freetype		little-cms2		zstd
genromfs		lz4

==> Casks
mch@mchdexuniji ~ % 
```

mrpollo commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -1,112 +1,83 @@
+   # Install required packages using pip3
+   python3 -m pip install --user pyserial empty toml numpy pandas jinja2 pyyaml pyros-genmsg packaging kconfiglib future jsonschema
+   # If this fails with a permissions error, your Python install is in a system path - use this command instead:
+   sudo -H python3 -m pip install --user pyserial empty toml numpy pandas jinja2 pyyaml pyros-genmsg packaging kconfiglib future jsonschema
```
@farhangnaderi can you please double check that the packages are not part of `Tools/setup/requirements.txt` if you find anything missing can you please submit a pull request to add them?

Replace this section with
```
# Install required packages using pip3
pip3 install -r Tools/setup/requirements.txt
# if this fails with a permissions error, try the following
sudo -H pip3 install -r Tools/setup/requirements.txt
```

mengchaoheng commented:
I am testing this part of the guide on a new macOS system. My architecture is Apple M1 Pro, the system is macOS 14.6 (23G80), and it is running well at the moment, but I really need gazebo classic, can you add this option to this part? @farhangnaderi 

My testing process::

1. `echo ulimit -S -n 2048 >> ~/.bashrc`

2. `alias pip3=/usr/bin/pip3` .  # Perhaps it is not necessary.

3.

```bash
brew tap PX4/px4
brew install px4-dev 

brew install px4-sim # error

brew link --overwrite --force arm-gcc-bin@13
```
4.python virtual environment

```bash
/usr/bin/python3 -m  venv px4_py

source px4_py/bin/activate


# delete --user
python3 -m pip install  pyserial empty toml numpy pandas jinja2 pyyaml pyros-genmsg packaging kconfiglib future jsonschema
``` 
5.

```bash
git clone https://github.com/PX4/PX4-Autopilot.git --recursive

cd PX4-Autopilot

make px4_sitl 

make px4_fmu-v5

```

Refer to [rolandash](https://club.mindpx.net/d/119-build-openmindpx-px4-on-apple-silicon-mac-with-native-toolchain/9
) to solve the problem:

```bash
Failed to import em: No module named 'em'

You may need to install it using:
    pip3 install --user empy

[6/1113] Building C object boards/px4/sitl/src/CMakeFiles/drivers_board.dir/sitl_led.c.o
ninja: build stopped: subcommand failed.
make: *** [px4_sitl] Error 1

```
Remember you are in python virtual environment:

```bash 

(px4_py) mch@mchdexuniji PX4-Autopilot % python3 -m pip install  empy==3.3.4 
Collecting empy==3.3.4
  Downloading empy-3.3.4.tar.gz (62 kB)
     |████████████████████████████████| 62 kB 126 kB/s 
Using legacy 'setup.py install' for empy, since package 'wheel' is not installed.
Installing collected packages: empy
    Running setup.py install for empy ... done
Successfully installed empy-3.3.4
WARNING: You are using pip version 21.2.4; however, version 25.2 is available.
You should consider upgrading via the '/Users/mch/px4_py/bin/python3 -m pip install --upgrade pip' command.
(px4_py) mch@mchdexuniji PX4-Autopilot % make px4_sitl                          
[0/1109] git submodule src/modules/uxrce_dds_client/Micro-XRCE-DDS-Client
[5/1109] git submodule src/modules/mavlink/mavlink
[9/1109] git submodule src/drivers/gps/devices
[11/1109] Generating Mavlink uAvionix: src/modules/mavlink/mavlink/message_definitions/v1.0/uAvionix.xml
WARNING: Failed to import lxml module etree. Are lxml, libxml2 and libxslt installed? XML validation will not be performed
[14/1109] Generating Mavlink development: src/modules/mavlink/mavlink/message_definitions/v1.0/development.xml
WARNING: Failed to import lxml module etree. Are lxml, libxml2 and libxslt installed? XML validation will not be performed
[99/1109] Linking CXX static library src/lib/slew_rate/libSlewRate.a
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: archive library: src/lib/slew_rate/libSlewRate.a the table of contents is empty (no object file members in the library define global symbols)
[1092/1109] Linking CXX executable bin/px4
ld: warning: ignoring duplicate libraries: '-lm', 'platforms/common/px4_work_queue/libpx4_work_queue.a', 'platforms/common/uORB/libuORB.a', 'platforms/posix/src/px4/common/libpx4_layer.a', 'platforms/posix/src/px4/common/lockstep_scheduler/liblockstep_scheduler.a', 'platforms/posix/src/px4/common/px4_daemon/libpx4_daemon.a', 'src/lib/parameters/libparameters.a', 'src/modules/dataman/libmodules__dataman.a', 'src/modules/navigator/MissionFeasibility/libmission_feasibility_checker.a', 'src/modules/navigator/libmodules__navigator.a'
[1109/1109] Linking CXX static library src/modules/ekf2/EKF/libecl_EKF.a
```

farhangnaderi commented:
> some ref. in [PX4/PX4-SITL_gazebo-classic#1021 (comment)](https://github.com/PX4/PX4-SITL_gazebo-classic/issues/1021#issuecomment-2078922511)
> 
> and https://club.mindpx.net/d/119-build-openmindpx-px4-on-apple-silicon-mac-with-native-toolchain/9

Keeping Gazebo classic is fine but there is no current maintainer for it and it is already EOL. SO I am not sure how sensible it will be for making it working.

hamishwillee commented:
@ryanjAA @farhangnaderi Is this as still stalled as it looks? What needs to happen? Someone want to buy me a mac?

farhangnaderi commented:
> @ryanjAA @farhangnaderi Is this as still stalled as it looks? What needs to happen? Someone want to buy me a mac?

I lost the track on this PR! 
We can wrap it up this week and merge hopefully.

farhangnaderi committed:
Cleanups

hamishwillee commented:
@mengchaoheng I apologize for not including you in the list of people I wanted feedback from in https://github.com/PX4/PX4-Autopilot/pull/25204#issuecomment-3383498623. You're right to feel a little underappreciated - you've contributed a lot here - which is why I wrote my next response. I suspect Ramon thought your response was an overreaction - I do too, but I get it.

From the following on discussion I think we all agree that it can be really frustrating to have our issues and feedback ignored. My only excuse is that I review PRs first as "biggest bang for buck" - by the end of doing that I have no time for anything else much.

Mostly I don't want this to stall. I am horrified this has take so long for us to fix.

mengchaoheng commented:
@mrpollo @hamishwillee problem with cmake can be found here: [rolandash](https://club.mindpx.net/d/119-build-openmindpx-px4-on-apple-silicon-mac-with-native-toolchain/9
)

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -4,109 +4,77 @@ The following instructions set up a PX4 development environment for macOS.
+1. Install Homebrew by following these [installation instructions](https://brew.sh).
 
-1. Enable more open files by appending the following line to the `~/.zshenv` file (creating it if necessary):
+2. Enable more open files by appending the following line to the `~/.bashrc` (or `~/.zshrc` for zsh)  file or creating it if necessary:
```
Not quite right:

```suggestion
2. Enable more files to be opened in the shell process by appending `echo ulimit -S -n 2048` to your `~/.bashrc` file (or `~/.zshrc` for zsh).
   To append the line using the terminal enter:
```

hamishwillee committed:
Update docs/en/dev_setup/dev_env_mac.md

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -4,109 +4,78 @@ The following instructions set up a PX4 development environment for macOS.
-   # Point pip3 to macOS system python 3 pip
-   alias pip3=/usr/bin/pip3
-   ```
+3. Enforce Python 3 by appending the following lines to `~/.bashrc`(or `~/.zshrc` for zsh):
````
What lines do you wish to append?

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -4,109 +4,78 @@ The following instructions set up a PX4 development environment for macOS.
    ```
 
-## Gazebo Classic Simulation
+2. ### Download PX4 Source Code
````
```suggestion
2. Download PX4 Source Code
```

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
````diff
@@ -4,109 +4,78 @@ The following instructions set up a PX4 development environment for macOS.
-To setup the environment for [Gazebo Classic](../sim_gazebo_classic/index.md) simulation:
+```sh
+git clone https://github.com/PX4/PX4-Autopilot.git --recursive
+```
````
````suggestion
   ```sh
   git clone https://github.com/PX4/PX4-Autopilot.git --recursive
   ```
````

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -4,109 +4,78 @@ The following instructions set up a PX4 development environment for macOS.
 
-1. Run the macOS setup script: `PX4-Autopilot/Tools/setup/macos.sh`
-   The easiest way to do this is to clone the PX4 source, and then run the script from the directory, as shown:
+   This script will install all required Homebrew packages, Python dependencies, and perform the necessary setup steps.
```
```suggestion
```

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -4,109 +4,78 @@ The following instructions set up a PX4 development environment for macOS.
-   :::
-
-1. To install SITL simulation with Gazebo Classic:
+   Use the [`macos.sh` setup script](https://github.com/PX4/PX4-Autopilot/blob/main/Tools/setup/macos.sh) to automatically install all development dependencies:
```
```suggestion
   Run the [`macos.sh` setup script](https://github.com/PX4/PX4-Autopilot/blob/main/Tools/setup/macos.sh) to automatically install all development dependencies (homebrew packages, python libraries, etc.):
```

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -4,109 +4,78 @@ The following instructions set up a PX4 development environment for macOS.
-   brew install --cask xquartz
-   brew install px4-sim-gazebo
+   git clone https://github.com/PX4/PX4-Autopilot.git --recursive   cd PX4-Autopilot
+   ./Tools/setup/macos.sh
```
@farhangnaderi there are a lot of very basic errors here. Did you build the docs and look at them?

```suggestion
   cd PX4-Autopilot
   ./Tools/setup/macos.sh
```

hamishwillee committed:
Apply suggestions from code review

hamishwillee committed:
Minor subedit and prettier

hamishwillee commented on docs/en/dev_setup/dev_env_mac.md:
```diff
@@ -4,109 +4,76 @@ The following instructions set up a PX4 development environment for macOS.
+## Gazebo Simulation
 
+Gazebo Harmonic simulation support is included with the `px4-sim` formula installed above.
+The simulation environment should be ready to use after the installation completes.
```
Is there a way to call macos.sh such that you don't install simulation tools?

hamishwillee reviewed (commented):
Thanks @farhangnaderi 
1. See https://github.com/PX4/PX4-Autopilot/pull/25204/files#r2496904317
2. So this will work for intel or Apple Silicon? What silicon has it been tested on/is it tested on?
3. What versions of Apple has it been tested on/is it expected to work on?

If it really is this simple, then this is wonderful.

farhangnaderi committed:
small correction

farhangnaderi committed:
cleanups gz harmonic brew formula

farhangnaderi commented:
> > 3. Error when running `make px4_sitl gz_x500`:
> > ```
> > ninja: error: unknown target 'gz_x500'
> > make: *** [px4_sitl] Error 1
> > ```
> 
> I will test it again, maybe tonight. The reason for the error might be that my virtual machine doesn't support starting Gazebo. I will test it on a brand-new computer.

@mengchaoheng you might want to check the last couple mentioned PRs/issues here as well.

mrpollo reviewed (dismissed):
Hey @farhangnaderi thanks for pushing so hard on this one, I will take over, please give me access to push to your branch

mrpollo committed:
fix(macos.sh): invert px4-sim install condition for --sim-tools

The condition checked if px4-sim WAS installed before running
brew install, meaning it would never install on a fresh system.
Add the missing negation so it installs when NOT already present.

Signed-off-by: Ramon Roche <mrpollo@gmail.com>

mrpollo committed:
docs: rewrite macOS dev environment setup guide

- Add Xcode Command Line Tools as prerequisite
- Default to ~/.zshrc (macOS default since Catalina)
- Explain why ulimit change is needed and why in startup file
- Add reminder to open new terminal after shell config changes
- Remove broken pip3 alias workaround
- Split git clone into clone + submodule update (canonical form)
- Recommend --sim-tools flag since first build uses gz_x500
- Document what macos.sh installs and its --reinstall flag
- Clarify Gazebo comes from --sim-tools / px4-sim formula
- Add XQuartz requirement for Gazebo display
- Add verification section with key tool checks and smoke test
- Remove outdated video guide comment block

Signed-off-by: Ramon Roche <mrpollo@gmail.com>

hamishwillee committed:
conventions

hamishwillee reviewed (commented):
@mrpollo Nice clean set of instructions. 

- https://github.com/PX4/PX4-Autopilot/pull/25957 is still open - doesn't that have to be fixed?
- seeing a clang error https://github.com/PX4/PX4-Autopilot/actions/runs/21935486908/job/63348658392?pr=25204 and the do

hamishwillee reviewed (approved):
I think this is good to go now then @mrpollo ?

Cannot wait!

farhangnaderi commented:
@mrpollo anyway I can revert it? 
I restored the branch.
